### PR TITLE
Fix repeat; rewrite multi-head attention layer

### LINKED ISSUE #157: `repeat` is not efficient on GPU
The ViT implementation contains  `repeat`, which will not work well for GPUs (due to scalar indexing)

https://github.com/FluxML/Metalhead.jl/blob/792076f1f86fdf713e2240cc0337c37ffd24fd1c/src/layers/embeddings.jl#L63

An alternative way to simulate repeat with broadcasting

https://github.com/avik-pal/Lux.jl/blob/49b69c7ed21ccde6fc8c07bacaf738de0b6fe497/lib/Boltz/src/vit.jl#L72

### PR DESCRIPTION
**Pull request by theabhirath**
Closes #157. Also borrows from the excellent implementation of `MHAttention` at https://github.com/avik-pal/Lux.jl/pull/29 to remove the NeuralAttentionlib dep, hopefully reducing time for `using Metalhead`

### CODE REVIEW

**Review by darsnack — CHANGES_REQUESTED** [review]

**Comment by darsnack** [comment]
It's a bit confusing...fields of structs should be `Int` and functions/constructor arguments should be `Integer`. You want the fields to be a concrete type for performance, you want arguments to be permissive and let the compiler auto-convert when needed.

**Comment by theabhirath** [comment]
> It's a bit confusing...fields of structs should be `Int` and functions/constructor arguments should be `Integer`. You want the fields to be a concrete type for performance, you want arguments to be permissive and let the compiler auto-convert when needed.

Gotcha, that makes sense

File: src/other/mlpmixer.jl [hunk 1]
```diff
@@ -132,7 +132,7 @@ return Chain(SkipConnection(Chain(Flux.Scale(planes),
                                   LayerScale(planes, λ),
                                   DropPath(drop_path_rate)), +),
              SkipConnection(Chain(Flux.Scale(planes),
-                                  mlp_layer(planes, Int(mlp_ratio * planes); dropout, activation),
+                                  mlp_layer(planes, Integer(mlp_ratio * planes); dropout, activation),
```
Comment (darsnack) [inline src/other/mlpmixer.jl:135]:
```suggestion
                                  mlp_layer(planes, Int(mlp_ratio * planes); dropout, activation),
```

**Review by darsnack — APPROVED** [review]
Sorry for the confusion, that was only meant to be a minor suggestion 😅.

**Comment by theabhirath** [comment]
That's quite alright, it helped me understand the rationale better! I've messed up the commits on this one a little bit though, so #159 supersedes this one

**Pull request closed without merging**